File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ComparableElmRequirement.java
package org.cqframework.cql.elm.requirements;

import static org.cqframework.cql.elm.evaluating.SimpleElmEvaluator.*;

import java.util.Iterator;
import org.cqframework.cql.cql2elm.tracking.Trackable;
import org.hl7.elm.r1.*;

public class ComparableElmRequirement {
    public ComparableElmRequirement(ElmRequirement requirement) {
        if (requirement == null) {
            throw new IllegalArgumentException("requirement is required");
        }

        this.requirement = requirement;
    }

    private ElmRequirement requirement;

    public ElmRequirement getRequirement() {
        return this.requirement;
    }

    @Override
    public int hashCode() {
        // Hashing only by the type/profile
        if (requirement.getElement() instanceof Retrieve) {
            Retrieve retrieve = (Retrieve) this.requirement.getElement();
            String typeUri = retrieve.getTemplateId() != null
                    ? retrieve.getTemplateId()
                    : (retrieve.getDataType() != null && retrieve.getDataType().getLocalPart() != null
                            ? retrieve.getDataType().getLocalPart()
                            : null);
            if (typeUri != null) {
                return typeUri.hashCode();
            }
        }

        return super.hashCode();
    }

    public static boolean codeFilterElementsEqual(CodeFilterElement left, CodeFilterElement right) {
        return stringsEqual(left.getProperty(), right.getProperty())
                && stringsEqual(left.getSearch(), right.getSearch())
                && stringsEqual(left.getComparator(), right.getComparator())
                && stringsEqual(left.getValueSetProperty(), right.getValueSetProperty())
                && codesEqual(left.getValue(), right.getValue());
    }

    public static boolean hasCodeFilter(Iterable<CodeFilterElement> left, CodeFilterElement right) {
        for (CodeFilterElement cfe : left) {
            if (codeFilterElementsEqual(cfe, right)) {
                return true;
            }
        }

        return false;
    }

    public static boolean codeFiltersEqual(Iterable<CodeFilterElement> left, Iterable<CodeFilterElement> right) {
        // TODO: Don't rely on order dependence here...
        Iterator<CodeFilterElement> leftIterator = left.iterator();
        Iterator<CodeFilterElement> rightIterator = right.iterator();

        while (leftIterator.hasNext()) {
            CodeFilterElement leftElement = leftIterator.next();

            if (!rightIterator.hasNext()) {
                return false;
            }

            CodeFilterElement rightElement = rightIterator.next();

            if (!codeFilterElementsEqual(leftElement, rightElement)) {
                return false;
            }
        }

        if (rightIterator.hasNext()) {
            return false;
        }

        return true;
    }

    public static boolean dateFilterElementsEqual(DateFilterElement left, DateFilterElement right) {
        return stringsEqual(left.getProperty(), right.getProperty())
                && stringsEqual(left.getLowProperty(), right.getLowProperty())
                && stringsEqual(left.getHighProperty(), right.getHighProperty())
                && stringsEqual(left.getSearch(), right.getSearch())
                && dateRangesEqual(left.getValue(), right.getValue());
    }

    public static boolean hasDateFilter(Iterable<DateFilterElement> left, DateFilterElement right) {
        for (DateFilterElement dfe : left) {
            if (dateFilterElementsEqual(dfe, right)) {
                return true;
            }
        }

        return false;
    }

    public static boolean dateFiltersEqual(Iterable<DateFilterElement> left, Iterable<DateFilterElement> right) {
        // TODO: Don't rely on order dependence here...
        Iterator<DateFilterElement> leftIterator = left.iterator();
        Iterator<DateFilterElement> rightIterator = right.iterator();

        while (leftIterator.hasNext()) {
            DateFilterElement leftElement = leftIterator.next();

            if (!rightIterator.hasNext()) {
                return false;
            }

            DateFilterElement rightElement = rightIterator.next();

            if (!dateFilterElementsEqual(leftElement, rightElement)) {
                return false;
            }
        }

        if (rightIterator.hasNext()) {
            return false;
        }

        return true;
    }

    // TODO: Handle types other than strings
    public static boolean otherFilterElementsEqual(OtherFilterElement left, OtherFilterElement right) {
        return stringsEqual(left.getProperty(), right.getProperty())
                && stringsEqual(left.getSearch(), right.getSearch())
                && stringsEqual(left.getComparator(), right.getComparator())
                && stringsEqual(left.getValue(), right.getValue());
    }

    public static boolean hasOtherFilter(Iterable<OtherFilterElement> left, OtherFilterElement right) {
        for (OtherFilterElement ofe : left) {
            if (otherFilterElementsEqual(ofe, right)) {
                return true;
            }
        }

        return false;
    }

    public static boolean otherFiltersEqual(Iterable<OtherFilterElement> left, Iterable<OtherFilterElement> right) {
        // TODO: Don't rely on order dependence here...
        Iterator<OtherFilterElement> leftIterator = left.iterator();
        Iterator<OtherFilterElement> rightIterator = right.iterator();

        while (leftIterator.hasNext()) {
            OtherFilterElement leftElement = leftIterator.next();

            if (!rightIterator.hasNext()) {
                return false;
            }

            OtherFilterElement rightElement = rightIterator.next();

            if (!otherFilterElementsEqual(leftElement, rightElement)) {
                return false;
            }
        }

        if (rightIterator.hasNext()) {
            return false;
        }

        return true;
    }

    public static boolean includeElementsEqual(IncludeElement left, IncludeElement right) {
        return left.getRelatedDataType() != null
                && right.getRelatedDataType() != null
                && stringsEqual(
                        left.getRelatedDataType().getNamespaceURI(),
                        right.getRelatedDataType().getNamespaceURI())
                && stringsEqual(
                        left.getRelatedDataType().getLocalPart(),
                        right.getRelatedDataType().getLocalPart())
                && stringsEqual(left.getRelatedProperty(), right.getRelatedProperty())
                && stringsEqual(left.getRelatedSearch(), right.getRelatedSearch());
    }

    public static boolean includeElementsEqual(Iterable<IncludeElement> left, Iterable<IncludeElement> right) {
        // TODO: Don't rely on order dependence here...
        Iterator<IncludeElement> leftIterator = left.iterator();
        Iterator<IncludeElement> rightIterator = right.iterator();

        while (leftIterator.hasNext()) {
            IncludeElement leftElement = leftIterator.next();

            if (!rightIterator.hasNext()) {
                return false;
            }

            IncludeElement rightElement = rightIterator.next();

            if (!includeElementsEqual(leftElement, rightElement)) {
                return false;
            }
        }

        if (rightIterator.hasNext()) {
            return false;
        }

        return true;
    }

    @Override
    public boolean equals(Object other) {
        if (other instanceof ComparableElmRequirement) {
            return requirementsEquivalent(requirement, ((ComparableElmRequirement) other).getRequirement());
        }

        return false;
    }

    public static boolean requirementsEquivalent(ElmRequirement left, ElmRequirement right) {
        Retrieve retrieve = (Retrieve) left.getElement();
        Retrieve otherRetrieve = (Retrieve) right.getElement();

        return retrieve.getDataType() != null
                && retrieve.getDataType().equals(otherRetrieve.getDataType())
                && stringsEqual(retrieve.getTemplateId(), otherRetrieve.getTemplateId())
                && stringsEqual(retrieve.getContext(), otherRetrieve.getContext())
                && stringsEqual(retrieve.getContextProperty(), otherRetrieve.getContextProperty())
                && stringsEqual(retrieve.getContextSearch(), otherRetrieve.getContextSearch())
                && stringsEqual(retrieve.getCodeProperty(), otherRetrieve.getCodeProperty())
                && stringsEqual(retrieve.getCodeSearch(), otherRetrieve.getCodeSearch())
                && stringsEqual(retrieve.getCodeComparator(), otherRetrieve.getCodeComparator())
                && stringsEqual(retrieve.getValueSetProperty(), otherRetrieve.getValueSetProperty())
                && codesEqual(retrieve.getCodes(), otherRetrieve.getCodes())
                && stringsEqual(retrieve.getDateProperty(), otherRetrieve.getDateProperty())
                && stringsEqual(retrieve.getDateLowProperty(), otherRetrieve.getDateLowProperty())
                && stringsEqual(retrieve.getDateHighProperty(), otherRetrieve.getDateHighProperty())
                && stringsEqual(retrieve.getDateSearch(), otherRetrieve.getDateSearch())
                && dateRangesEqual(retrieve.getDateRange(), otherRetrieve.getDateRange())
                && stringsEqual(retrieve.getIdProperty(), otherRetrieve.getIdProperty())
                && stringsEqual(retrieve.getIdSearch(), otherRetrieve.getIdSearch())
                && stringsEqual(retrieve.getId(), otherRetrieve.getId())
                && codeFiltersEqual(retrieve.getCodeFilter(), otherRetrieve.getCodeFilter())
                && dateFiltersEqual(retrieve.getDateFilter(), otherRetrieve.getDateFilter())
                && otherFiltersEqual(retrieve.getOtherFilter(), otherRetrieve.getOtherFilter())
                // TODO: support for collapsing includes
                && stringsEqual(retrieve.getIncludedIn(), otherRetrieve.getIncludedIn())
                && includeElementsEqual(retrieve.getInclude(), otherRetrieve.getInclude());
    }

    private static boolean hasInclude(Retrieve retrieve, IncludeElement includeElement) {
        for (IncludeElement e : retrieve.getInclude()) {
            if (includeElementsEqual(e, includeElement)) {
                return true;
            }
        }

        return false;
    }

    public static ElmRequirement mergeRequirements(ElmRequirement existing, ElmRequirement required) {
        if (existing instanceof ElmDataRequirement) {
            if (required instanceof ElmDataRequirement) {
                if (existing.getElement() instanceof Retrieve) {
                    if (required.getElement() instanceof Retrieve) {
                        Retrieve existingRetrieve = (Retrieve) existing.getElement();
                        Retrieve requiredRetrieve = (Retrieve) required.getElement();
                        Retrieve newRetrieve = ElmCloner.clone(existingRetrieve);

                        var trackbacks = Trackable.INSTANCE.getTrackbacks(requiredRetrieve);
                        var newTrackbacks = Trackable.INSTANCE.getTrackbacks(newRetrieve);
                        newTrackbacks.addAll(trackbacks);

                        ElmDataRequirement newRequirement =
                                new ElmDataRequirement(existing.getLibraryIdentifier(), newRetrieve);
                        if (((ElmDataRequirement) existing).getProperties() != null) {
                            for (Property property : ((ElmDataRequirement) existing).getProperties()) {
                                newRequirement.addProperty(property);
                            }
                        }

                        // Merge mustSupport
                        if (((ElmDataRequirement) required).getProperties() != null) {
                            for (Property property : ((ElmDataRequirement) required).getProperties()) {
                                newRequirement.addProperty(property);
                            }
                        }

                        // Merge includes
                        for (IncludeElement includeElement : requiredRetrieve.getInclude()) {
                            if (!hasInclude(newRetrieve, includeElement)) {
                                newRetrieve.getInclude().add(ElmCloner.clone(includeElement));
                            }
                        }

                        return newRequirement;
                    }
                }
            }
        }

        return existing;
    }
}

File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ComparableElmRequirement.kt
package org.cqframework.cql.elm.requirements

import org.cqframework.cql.cql2elm.tracking.Trackable.trackbacks
import org.cqframework.cql.elm.evaluating.SimpleElmEvaluator.codesEqual
import org.cqframework.cql.elm.evaluating.SimpleElmEvaluator.dateRangesEqual
import org.cqframework.cql.elm.evaluating.SimpleElmEvaluator.stringsEqual
import org.hl7.elm.r1.CodeFilterElement
import org.hl7.elm.r1.DateFilterElement
import org.hl7.elm.r1.IncludeElement
import org.hl7.elm.r1.OtherFilterElement
import org.hl7.elm.r1.Retrieve

class ComparableElmRequirement(val requirement: ElmRequirement) {

    override fun hashCode(): Int {
        // Hashing only by the type/profile
        if (requirement.getElement() is Retrieve) {
            val retrieve = this.requirement.getElement() as Retrieve
            val typeUri =
                if (retrieve.templateId != null) retrieve.templateId
                else
                    (if (retrieve.dataType != null && retrieve.dataType!!.localPart != null)
                        retrieve.dataType!!.localPart
                    else null)
            if (typeUri != null) {
                return typeUri.hashCode()
            }
        }

        return super.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (other is ComparableElmRequirement) {
            return requirementsEquivalent(requirement, other.requirement)
        }

        return false
    }

    companion object {
        @JvmStatic
        fun codeFilterElementsEqual(left: CodeFilterElement, right: CodeFilterElement): Boolean {
            return stringsEqual(left.property, right.property) &&
                stringsEqual(left.search, right.search) &&
                stringsEqual(left.comparator, right.comparator) &&
                stringsEqual(left.valueSetProperty, right.valueSetProperty) &&
                codesEqual(left.value, right.value)
        }

        @JvmStatic
        fun hasCodeFilter(left: Iterable<CodeFilterElement>, right: CodeFilterElement): Boolean {
            for (cfe in left) {
                if (codeFilterElementsEqual(cfe, right)) {
                    return true
                }
            }

            return false
        }

        @JvmStatic
        fun codeFiltersEqual(
            left: Iterable<CodeFilterElement>,
            right: Iterable<CodeFilterElement>,
        ): Boolean {
            // TODO: Don't rely on order dependence here...
            val leftIterator = left.iterator()
            val rightIterator = right.iterator()

            while (leftIterator.hasNext()) {
                val leftElement = leftIterator.next()

                if (!rightIterator.hasNext()) {
                    return false
                }

                val rightElement = rightIterator.next()

                if (!codeFilterElementsEqual(leftElement, rightElement)) {
                    return false
                }
            }

            return !rightIterator.hasNext()
        }

        @JvmStatic
        fun dateFilterElementsEqual(left: DateFilterElement, right: DateFilterElement): Boolean {
            return stringsEqual(left.property, right.property) &&
                stringsEqual(left.lowProperty, right.lowProperty) &&
                stringsEqual(left.highProperty, right.highProperty) &&
                stringsEqual(left.search, right.search) &&
                dateRangesEqual(left.value, right.value)
        }

        @JvmStatic
        fun hasDateFilter(left: Iterable<DateFilterElement>, right: DateFilterElement): Boolean {
            for (dfe in left) {
                if (dateFilterElementsEqual(dfe, right)) {
                    return true
                }
            }

            return false
        }

        @JvmStatic
        fun dateFiltersEqual(
            left: Iterable<DateFilterElement>,
            right: Iterable<DateFilterElement>,
        ): Boolean {
            // TODO: Don't rely on order dependence here...
            val leftIterator = left.iterator()
            val rightIterator = right.iterator()

            while (leftIterator.hasNext()) {
                val leftElement = leftIterator.next()

                if (!rightIterator.hasNext()) {
                    return false
                }

                val rightElement = rightIterator.next()

                if (!dateFilterElementsEqual(leftElement, rightElement)) {
                    return false
                }
            }

            return !rightIterator.hasNext()
        }

        @JvmStatic
        // TODO: Handle types other than strings
        fun otherFilterElementsEqual(left: OtherFilterElement, right: OtherFilterElement): Boolean {
            return stringsEqual(left.property, right.property) &&
                stringsEqual(left.search, right.search) &&
                stringsEqual(left.comparator, right.comparator) &&
                stringsEqual(left.value, right.value)
        }

        @JvmStatic
        fun hasOtherFilter(left: Iterable<OtherFilterElement>, right: OtherFilterElement): Boolean {
            for (ofe in left) {
                if (otherFilterElementsEqual(ofe, right)) {
                    return true
                }
            }

            return false
        }

        @JvmStatic
        fun otherFiltersEqual(
            left: Iterable<OtherFilterElement>,
            right: Iterable<OtherFilterElement>,
        ): Boolean {
            // TODO: Don't rely on order dependence here...
            val leftIterator = left.iterator()
            val rightIterator = right.iterator()

            while (leftIterator.hasNext()) {
                val leftElement = leftIterator.next()

                if (!rightIterator.hasNext()) {
                    return false
                }

                val rightElement = rightIterator.next()

                if (!otherFilterElementsEqual(leftElement, rightElement)) {
                    return false
                }
            }

            return !rightIterator.hasNext()
        }

        @JvmStatic
        fun includeElementsEqual(left: IncludeElement, right: IncludeElement): Boolean {
            return left.relatedDataType != null &&
                right.relatedDataType != null &&
                stringsEqual(
                    left.relatedDataType!!.namespaceURI,
                    right.relatedDataType!!.namespaceURI,
                ) &&
                stringsEqual(left.relatedDataType!!.localPart, right.relatedDataType!!.localPart) &&
                stringsEqual(left.relatedProperty, right.relatedProperty) &&
                stringsEqual(left.relatedSearch, right.relatedSearch)
        }

        @JvmStatic
        fun includeElementsEqual(
            left: Iterable<IncludeElement>,
            right: Iterable<IncludeElement>,
        ): Boolean {
            // TODO: Don't rely on order dependence here...
            val leftIterator = left.iterator()
            val rightIterator = right.iterator()

            while (leftIterator.hasNext()) {
                val leftElement = leftIterator.next()

                if (!rightIterator.hasNext()) {
                    return false
                }

                val rightElement = rightIterator.next()

                if (!includeElementsEqual(leftElement, rightElement)) {
                    return false
                }
            }

            return !rightIterator.hasNext()
        }

        @JvmStatic
        fun requirementsEquivalent(left: ElmRequirement, right: ElmRequirement): Boolean {
            val retrieve = left.getElement() as Retrieve
            val otherRetrieve = right.getElement() as Retrieve

            return retrieve.dataType != null &&
                retrieve.dataType!! == otherRetrieve.dataType &&
                stringsEqual(retrieve.templateId, otherRetrieve.templateId) &&
                stringsEqual(retrieve.context, otherRetrieve.context) &&
                stringsEqual(retrieve.contextProperty, otherRetrieve.contextProperty) &&
                stringsEqual(retrieve.contextSearch, otherRetrieve.contextSearch) &&
                stringsEqual(retrieve.codeProperty, otherRetrieve.codeProperty) &&
                stringsEqual(retrieve.codeSearch, otherRetrieve.codeSearch) &&
                stringsEqual(retrieve.codeComparator, otherRetrieve.codeComparator) &&
                stringsEqual(retrieve.valueSetProperty, otherRetrieve.valueSetProperty) &&
                codesEqual(retrieve.codes, otherRetrieve.codes) &&
                stringsEqual(retrieve.dateProperty, otherRetrieve.dateProperty) &&
                stringsEqual(retrieve.dateLowProperty, otherRetrieve.dateLowProperty) &&
                stringsEqual(retrieve.dateHighProperty, otherRetrieve.dateHighProperty) &&
                stringsEqual(retrieve.dateSearch, otherRetrieve.dateSearch) &&
                dateRangesEqual(retrieve.dateRange, otherRetrieve.dateRange) &&
                stringsEqual(retrieve.idProperty, otherRetrieve.idProperty) &&
                stringsEqual(retrieve.idSearch, otherRetrieve.idSearch) &&
                stringsEqual(retrieve.id, otherRetrieve.id) &&
                codeFiltersEqual(retrieve.codeFilter, otherRetrieve.codeFilter) &&
                dateFiltersEqual(retrieve.dateFilter, otherRetrieve.dateFilter) &&
                otherFiltersEqual(
                    retrieve.otherFilter,
                    otherRetrieve.otherFilter,
                ) // TODO: support for collapsing includes
                &&
                stringsEqual(retrieve.includedIn, otherRetrieve.includedIn) &&
                includeElementsEqual(retrieve.include, otherRetrieve.include)
        }

        @JvmStatic
        private fun hasInclude(retrieve: Retrieve, includeElement: IncludeElement): Boolean {
            for (e in retrieve.include) {
                if (includeElementsEqual(e, includeElement)) {
                    return true
                }
            }

            return false
        }

        @JvmStatic
        fun mergeRequirements(existing: ElmRequirement, required: ElmRequirement): ElmRequirement {
            if (existing is ElmDataRequirement) {
                if (required is ElmDataRequirement) {
                    if (existing.getElement() is Retrieve) {
                        if (required.getElement() is Retrieve) {
                            val existingRetrieve = existing.getElement() as Retrieve
                            val requiredRetrieve = required.getElement() as Retrieve
                            val newRetrieve = ElmCloner.clone(existingRetrieve)

                            val trackbacks = requiredRetrieve.trackbacks
                            val newTrackbacks = newRetrieve.trackbacks
                            newTrackbacks.addAll(trackbacks)

                            val newRequirement =
                                ElmDataRequirement(existing.getLibraryIdentifier(), newRetrieve)
                            if (existing.properties != null) {
                                for (property in existing.properties) {
                                    newRequirement.addProperty(property)
                                }
                            }

                            // Merge mustSupport
                            if (required.properties != null) {
                                for (property in required.properties) {
                                    newRequirement.addProperty(property)
                                }
                            }

                            // Merge includes
                            for (includeElement in requiredRetrieve.include) {
                                if (!hasInclude(newRetrieve, includeElement)) {
                                    newRetrieve.include.add(ElmCloner.clone(includeElement))
                                }
                            }

                            return newRequirement
                        }
                    }
                }
            }

            return existing
        }
    }
}
